feat(connext): removeInvoice

sangaman opened the pull request:
This implements the new functionality to cancel incoming hash lock transfers in Connext, automatically freeing any HTLCs.

Closes #1730.

This is still a draft as it needs testing with the newest connext versions. I also believe this might require some modifications to the rest-api-client to ensure the `resolveCondition` call is happening properly, and it might make sense to create a separate endpoint for hashlock cancellations only such as `/hashlock-cancel`. I can open a PR for that the way I created `/withdraw`

erkarl commented:
I think a separate endpoint for cancellation makes sense. Let's see what the API will look like when the functionality is released from connext side.

sangaman commented:
Depending on https://github.com/connext/rest-api-client/pull/60 for a new endpoint to use in this PR.

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -400,7 +400,17 @@ class ConnextClient extends SwapClient {
+      // resolve a hashlock with a paymentId but no preimage to cancel it
+      await this.sendRequest('/hashlock-resolve', 'POST', {
+        assetId: expectedIncomingTransfer.tokenAddress,
+        paymentId: rHash,
```
This won't work because `paymentId` is `hash(preimage + assetId)`. I tested https://github.com/connext/rest-api-client/pull/60 manually using the paymentId and the receiver was able to cancel the transfer.

sangaman replied:
Thanks, this wasn't completely clear to me. Do you happen to have an example or instructions of how exactly this hash is calculated? Do I append the token address to the preimage, decode from hex to bytes, and then sha256 the bytes?

erkarl replied:
As discussed offline I think this would be easier if we could get the `paymentId` from the response of `/hashlock-status/$HASH/$ASSET_ID`

sangaman replied:
I implemented a solution to save the `paymentId` from the created event. It won't persist through an xud restart but it'll allow us to cancel invoices for swaps where xud stays online (which should be most cases).

raladev commented:
got 400 err for connect request:
![conn_err](https://user-images.githubusercontent.com/29906866/89131937-26330d00-d519-11ea-8bc6-bc104e9a0cd4.png)

sangaman commented:
> got 400 err for connect request:

Is this happening consistently for you? I can't think of how the changes in this PR would affect the `/connect` call.

erkarl commented:
> > got 400 err for connect request:
> 
> Is this happening consistently for you? I can't think of how the changes in this PR would affect the `/connect` call.

I also managed to reproduce this. Think it's related to: https://github.com/ExchangeUnion/xud/pull/1743

Can you please try rebasing this branch?

sangaman commented:
> I also managed to reproduce this. Think it's related to: #1743
> 
> Can you please try rebasing this branch?

Ahh that makes a lot of sense, rebased and thanks for testing it.

erkarl commented:
I tested the following scenario:

1. A (maker) created sell ETH/BTC order
2. B (taker) created buy ETH/BTC order (with modified xud code to NOT unlock the ETH payment)
3. Deal timed out
4. ERROR: B did not cancel ETH payment
5. Once B cancels the ETH payment, A should also cancel the BTC invoice

Taker:
```
05/08/2020 20:44:52.746 [SWAPS] debug: New deal: {"takerCltvDelta":200,"rHash":"8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","orderId":"7f2118e0-d75c-11ea-b74e-61489408d92b","pairId":"ETH/BTC","proposedQuantity":100000,"rPreimage":"8ac7d24f4cccbbd3ce8e71d6077224acfd4a298b194822de4db7684672dcad16","takerCurrency":"ETH","makerCurrency":"BTC","takerAmount":100000,"makerAmount":100000,"takerUnits":1000000000000000,"makerUnits":100000,"destination":"02e4f47f06b952353636e6975bd0b1b496f40ac90992d9ebce05849f0c7e642839","peerPubKey":"02a9857a082c5a269a36a2bda029b2d88ed1c179104e26af000314c27eb5a2b82a","localId":"8341c4b0-d75c-11ea-a525-f5d88a98f23d","price":1,"isBuy":false,"phase":0,"state":0,"role":0,"createTime":1596660292745}
05/08/2020 20:44:52.747 [SWAPS] debug: Requesting deal: {"takerCltvDelta":200,"rHash":"8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","orderId":"7f2118e0-d75c-11ea-b74e-61489408d92b","pairId":"ETH/BTC","proposedQuantity":100000,"rPreimage":"8ac7d24f4cccbbd3ce8e71d6077224acfd4a298b194822de4db7684672dcad16","takerCurrency":"ETH","makerCurrency":"BTC","takerAmount":100000,"makerAmount":100000,"takerUnits":1000000000000000,"makerUnits":100000,"destination":"02e4f47f06b952353636e6975bd0b1b496f40ac90992d9ebce05849f0c7e642839","peerPubKey":"02a9857a082c5a269a36a2bda029b2d88ed1c179104e26af000314c27eb5a2b82a","localId":"8341c4b0-d75c-11ea-a525-f5d88a98f23d","price":1,"isBuy":false,"phase":0,"state":0,"role":0,"createTime":1596660292745}
05/08/2020 20:44:52.762 [P2P] debug: received swapAccepted from 02a9857a082c5a269a36a2bda029b2d88ed1c179104e26af000314c27eb5a2b82a (ErrorEnter): {"rHash":"8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","quantity":100000,"makerCltvDelta":166}
05/08/2020 20:44:52.763 [SWAPS] debug: Setting SendingPayment phase for deal 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:44:52.768 [LND-BTC] debug: sending payment of 100000 with hash 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730 to 02e4f47f06b952353636e6975bd0b1b496f40ac90992d9ebce05849f0c7e642839
05/08/2020 20:44:54.294 [HTTP] debug: http server request json: {"id":"d79c6fd9-96da-45b4-9a7f-d2eea836509e","data":{"amount":{"_hex":"0x038d7ea4c68000","_isBigNumber":true},"appIdentityHash":"0xbc8d037636afdf807c785a2dd62168cd5ed408ad580c4a0b29cf62946394e486","assetId":"0x0000000000000000000000000000000000000000","meta":{"recipient":"indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6","sender":"indra81wwkDhzSpCjy9NZxCaFZZUBdzUA1qLuwoSoMLMmkevoAgoghh","paymentId":"0x093c6e3b3a849bdf9753565fb45bc77e19f773e07af142e65380b0a71188aa77","timelock":"200"},"sender":"indra81wwkDhzSpCjy9NZxCaFZZUBdzUA1qLuwoSoMLMmkevoAgoghh","transferMeta":{"lockHash":"0x8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","expiry":{"_hex":"0x013fb7","_isBigNumber":true},"timelock":"200"},"type":"HashLockTransferApp","paymentId":"0x093c6e3b3a849bdf9753565fb45bc77e19f773e07af142e65380b0a71188aa77","recipient":"indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6"}}
05/08/2020 20:44:54.295 [CONNEXT] debug: accepting incoming transfer with rHash: 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730, units: 1000000000000000, timelock 200 and currency ETH
05/08/2020 20:44:54.373 [LND-BTC] debug: new channel maximum outbound capacity: 8403538
05/08/2020 20:44:54.373 [LND-BTC] debug: new channel inbound capacity: 7165262
05/08/2020 20:44:54.373 [LND-BTC] debug: new channel total outbound capacity: 8403538
05/08/2020 20:46:22.169 [P2P] debug: received swapFailed due to UnknownError from 02a9857a082c5a269a36a2bda029b2d88ed1c179104e26af000314c27eb5a2b82a (ErrorEnter): {"rHash":"8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","errorMessage":"Timeout has occurred","failureReason":12}
05/08/2020 20:46:22.170 [SWAPS] debug: deal 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730 failed due to UnknownError: Timeout has occurred
05/08/2020 20:46:22.170 [SWAPS] warn: swap failed due to unknown error: Timeout has occurred
05/08/2020 20:46:22.171 [CONNEXT] warn: could not find expected incoming transfer with hash 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:46:22.172 [ORDERBOOK] error: swap between orders 7f2118e0-d75c-11ea-b74e-61489408d92b & 8341c4b0-d75c-11ea-a525-f5d88a98f23d failed due to UnknownError
05/08/2020 20:46:22.172 [ORDERBOOK] warn: swap for 100000 failed during order matching due to UnknownError, will repeat matching routine for failed quantity
05/08/2020 20:46:22.173 [ORDERBOOK] debug: repeating matching routine for 8341c4b0-d75c-11ea-a525-f5d88a98f23d for failed quantity of 100000
05/08/2020 20:46:22.174 [ORDERBOOK] debug: placeOrder max time exceeded. order ({"pairId":"ETH/BTC","price":null,"quantity":100000,"isBuy":true,"localId":"8341c4b0-d75c-11ea-a525-f5d88a98f23d","id":"8341c4b0-d75c-11ea-a525-f5d88a98f23d","initialQuantity":100000,"hold":0,"createdAt":1596660292731}) won't be fully matched
```

Maker:
```
05/08/2020 20:44:45.805 [RPC] debug: received call /xudrpc.Xud/PlaceOrder
05/08/2020 20:44:52.746 [P2P] debug: received swapRequest from 032aa721969c80f79a77fe79c641277637edc6616c4764fe3de00ff49328f27941 (JokeFever): {"proposedQuantity":100000,"pairId":"ETH/BTC","orderId":"7f2118e0-d75c-11ea-b74e-61489408d92b","rHash":"8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","takerCltvDelta":200}
05/08/2020 20:44:52.747 [ORDERBOOK] debug: added hold of 100000 on order 7f2118e0-d75c-11ea-b74e-61489408d92b
05/08/2020 20:44:52.749 [SWAPS] debug: trying to accept deal: {"quantity":100000,"localId":"7f2118e0-d75c-11ea-b74e-61489408d92b","price":1,"isBuy":false} from xudPubKey: 032aa721969c80f79a77fe79c641277637edc6616c4764fe3de00ff49328f27941
05/08/2020 20:44:52.750 [SWAPS] debug: New deal: {"rHash":"8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","pairId":"ETH/BTC","proposedQuantity":100000,"orderId":"7f2118e0-d75c-11ea-b74e-61489408d92b","price":1,"isBuy":false,"quantity":100000,"makerAmount":100000,"takerAmount":100000,"makerCurrency":"BTC","takerCurrency":"ETH","makerUnits":100000,"takerUnits":1000000000000000,"takerCltvDelta":200,"takerPubKey":"indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6","destination":"indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6","peerPubKey":"032aa721969c80f79a77fe79c641277637edc6616c4764fe3de00ff49328f27941","localId":"7f2118e0-d75c-11ea-b74e-61489408d92b","phase":0,"state":0,"role":1,"createTime":1596660292750}
05/08/2020 20:44:52.750 [SWAPS] debug: got ETH block height of 1
05/08/2020 20:44:52.751 [SWAPS] debug: found route to taker with total lock duration of 100 ETH blocks (~0h)
05/08/2020 20:44:52.751 [SWAPS] debug: calculated lock buffer for first leg: 22 BTC blocks (~4h)
05/08/2020 20:44:52.752 [SWAPS] debug: lock delta for final hop to maker: 166 BTC blocks (~28h)
05/08/2020 20:44:52.755 [LND-BTC] debug: added invoice of 100000 for 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730 with cltvExpiry 166
05/08/2020 20:44:52.756 [SWAPS] debug: Setting SwapAccepted phase for deal 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:44:52.756 [SWAPS] debug: sending swap accepted packet: {"rHash":"8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","makerCltvDelta":166,"quantity":100000} to peer: 032aa721969c80f79a77fe79c641277637edc6616c4764fe3de00ff49328f27941
05/08/2020 20:44:52.899 [LND-BTC] debug: accepted htlc for invoice 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:44:52.899 [SWAPS] debug: Executing maker code to resolve hash
05/08/2020 20:44:52.899 [SWAPS] debug: Setting SendingPayment phase for deal 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:44:52.903 [CONNEXT] debug: sending payment of 1000000000000000 with hash 0x8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730 to indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6
05/08/2020 20:44:54.155 [HTTP] debug: http server request json: {"id":"bb123683-1d30-4aa9-a91d-de4d015e73c5","data":{"amount":{"_hex":"0x038d7ea4c68000","_isBigNumber":true},"appIdentityHash":"0x1b2ee5e3a8d40d7b1d7ae88b0f9299b84aad64c0d37850467d02fb4a78b7abd5","assetId":"0x0000000000000000000000000000000000000000","meta":{"recipient":"indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6","sender":"indra81wwkDhzSpCjy9NZxCaFZZUBdzUA1qLuwoSoMLMmkevoAgoghh","paymentId":"0x093c6e3b3a849bdf9753565fb45bc77e19f773e07af142e65380b0a71188aa77","timelock":"200"},"sender":"indra81wwkDhzSpCjy9NZxCaFZZUBdzUA1qLuwoSoMLMmkevoAgoghh","transferMeta":{"lockHash":"0x8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","expiry":{"_hex":"0x01401b","_isBigNumber":true},"timelock":"200"},"type":"HashLockTransferApp","paymentId":"0x093c6e3b3a849bdf9753565fb45bc77e19f773e07af142e65380b0a71188aa77","recipient":"indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6"}}
05/08/2020 20:44:54.155 [CONNEXT] warn: received unexpected incoming transfer created event with rHash 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:44:54.254 [HTTP] debug: http server request json: {"id":"bb123683-1d30-4aa9-a91d-de4d015e73c5","data":{"type":"HashLockTransferApp","amount":{"_hex":"0x038d7ea4c68000","_isBigNumber":true},"appIdentityHash":"0x1b2ee5e3a8d40d7b1d7ae88b0f9299b84aad64c0d37850467d02fb4a78b7abd5","assetId":"0x0000000000000000000000000000000000000000","sender":"indra81wwkDhzSpCjy9NZxCaFZZUBdzUA1qLuwoSoMLMmkevoAgoghh","meta":{"recipient":"indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6","sender":"indra81wwkDhzSpCjy9NZxCaFZZUBdzUA1qLuwoSoMLMmkevoAgoghh","paymentId":"0x093c6e3b3a849bdf9753565fb45bc77e19f773e07af142e65380b0a71188aa77","timelock":"200"},"paymentId":"0x093c6e3b3a849bdf9753565fb45bc77e19f773e07af142e65380b0a71188aa77","recipient":"indra5UtRK8PzK6iu7VSo2xpNhyoGbMX2xQtHvVjgEDCx6M2dkYbzE6","transferMeta":{"expiry":{"_hex":"0x01401b","_isBigNumber":true},"timelock":"200","lockHash":"0x8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730"}}}
05/08/2020 20:44:54.254 [CONNEXT] warn: received unexpected incoming transfer created event with rHash 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:44:55.832 [LND-BTC] debug: new channel maximum outbound capacity: 7165262
05/08/2020 20:44:55.832 [LND-BTC] debug: new channel inbound capacity: 8403538
05/08/2020 20:44:55.833 [LND-BTC] debug: new channel total outbound capacity: 7165262
05/08/2020 20:46:21.905 [SWAPS] debug: sendPayment in resolveHash failed due to Timeout has occurred
05/08/2020 20:46:22.034 [SWAPS] debug: deal 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730 failed due to UnknownError (ETH): Timeout has occurred
05/08/2020 20:46:22.034 [SWAPS] warn: swap failed due to unknown error: Timeout has occurred
05/08/2020 20:46:22.046 [SWAPS-RECOVERY] info: recovering swap deal 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:46:22.046 [SWAPS-RECOVERY] debug: checking outgoing payment status for swap 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:46:22.164 [SWAPS-RECOVERY] debug: swap for 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730 still has pending payments and will be monitored
05/08/2020 20:46:22.165 [ORDERBOOK] debug: removed hold of 100000 on order 7f2118e0-d75c-11ea-b74e-61489408d92b
05/08/2020 20:46:22.166 [SWAPS] debug: Sending UnknownError error to peer: {"rHash":"8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730","failureReason":12,"errorMessage":"Timeout has occurred"}
05/08/2020 20:46:22.167 [SWAPS] error: could not resolve hash for deal 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730: 7.9 - Timeout has occurred
05/08/2020 20:47:49.651 [SWAPS-RECOVERY] debug: checking outgoing payment status for swap 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730
05/08/2020 20:47:49.779 [SWAPS-RECOVERY] debug: swap for 8a08eb4fcf325f1f19249a4c9d8864ba08d23755de802a4bfe11b8312b7f7730 still has pending payments and will be monitored
```

raladev commented:
results In short - I cant find reliable legitimate way to get cancel from counterparty.

- tried connext shutdown - fund was returned back from the node to sender because of 60 timeout
- tried xud shutdown + connext shutodwn - it does not work because i lost paymentId ( got cant find a incoming tranfer msg) and funds has not been returned to sender
- tried lndbtc shutdown during swap and upping it back -  got cant find a incoming tranfer msg and funds has not been returned to sender

IMO cancel should work in third case, so:

Steps:
1. Place 8 sell BTC/ETH orders as a maker
2. Send 1 buy order as a maker that will be matched with all 8 orders ( [Screenshot from 2020-08-05 22-33-15](https://user-images.githubusercontent.com/29906866/89463132-24668500-d777-11ea-876a-7ed81b43e692.png)
)
3. Wait first successful match in xud log
4. docker rm -f simnet_lndbtc_1   (after this channel of taker that was connected to simnet node was closed with no reason)
5. Check taker ETH balance -> Full amount of 8 orders was blocked  ( [Screenshot from 2020-08-05 22-33-15](https://user-images.githubusercontent.com/29906866/89463132-24668500-d777-11ea-876a-7ed81b43e692.png)(0.0008*8+0.0007*7+0.0006*6+0.0005*5+0.0004*4+0.0003*3+0.0002*2+0.0001*1=0.0204 ETH)
5. Wait first recovery attempt from maker -> all recovery attempts were failed because of disabled BTC node
6. Return lndbtc node back
7. Wait second recovery attempt

Actual result:

1. 1 successful recovery for released ETH payment (because 2 swaps were processed before disablin lndbtc, but only one was processed by xud. So, recovery solved this problem). As a result we have 2 successful swaps of 8 - for 0.0001(0.0001*1) ETH and for 0.0025(0.0005*5).

2. 
```
swap for a7bbc94ac3bac589488dba7ceeed3cada3ba763a613a100776eee65624d0af13 still has pending payments and will be monitored
swap for 2b48974858ec68b34d174c1580986debf5e4d654948bfd84c9b661cf7d42a73d still has pending payments and will be monitored
```
a7bbc94ac3bac589488dba7ceeed3cada3ba763a613a100776eee65624d0af13 swap is for 0.0003*3 ETH
2b48974858ec68b34d174c1580986debf5e4d654948bfd84c9b661cf7d42a73d swap is for 0.0006*6 ETH

3.
```
could not find expected incoming transfer with hash eaaef273e2bdd497962829e3e36bf84a169bf07cfa485fbd9acf3ac3943dd426
could not find expected incoming transfer with hash 1dc7108f3461f4844bfccb4cc9377447b0a437cf10d7ef030122bf15442bae51
could not find expected incoming transfer with hash 63b53468480f0eed9ca37789fc931d0ec4eb4a8e7eff3760371f59bfd50abe4a
could not find expected incoming transfer with hash 5398e3d05d3a8aa72ff1514dc36bab0e25e9f2c0ccca8142753edee4ea93d0b9
```
eaaef273e2bdd497962829e3e36bf84a169bf07cfa485fbd9acf3ac3943dd426 swap is for 0.0007*7 ETH
1dc7108f3461f4844bfccb4cc9377447b0a437cf10d7ef030122bf15442bae51 swap is for 0.0008*8 ETH
63b53468480f0eed9ca37789fc931d0ec4eb4a8e7eff3760371f59bfd50abe4a swap is for 0.0004*4 ETH 
5398e3d05d3a8aa72ff1514dc36bab0e25e9f2c0ccca8142753edee4ea93d0b9 swap is for 0.0002*2 ETH

4. taker still has 0.0204 of blocked amount, so nothing was returned back

[connext_taker_nocancel.log](https://github.com/ExchangeUnion/xud/files/5031033/connext_taker_nocancel.log)
[xud_taker_nocancel.log](https://github.com/ExchangeUnion/xud/files/5031034/xud_taker_nocancel.log)
[connext_maker_nocancel.log](https://github.com/ExchangeUnion/xud/files/5031035/connext_maker_nocancel.log)
[xud_maker_nocancel.log](https://github.com/ExchangeUnion/xud/files/5031036/xud_maker_nocancel.log)


5. Wait third attempt of recovery ->
Got only this and nothing about other 6 swaps. Also i checked balance of ETH for the taker. Nothing was returned back, he still has 0.0204 blocked ETH.
```
swap for a7bbc94ac3bac589488dba7ceeed3cada3ba763a613a100776eee65624d0af13 still has pending payments and will be monitored
swap for 2b48974858ec68b34d174c1580986debf5e4d654948bfd84c9b661cf7d42a73d still has pending payments and will be monitored
```

6. VERY IMPORTANT NOTE:
maker's lndbtc has only one pending htlc (for 30001 satoshi) - one of the pending swaps
[Screenshot from 2020-08-05 23-58-37](https://user-images.githubusercontent.com/29906866/89463462-a656ae00-d777-11ea-96c4-4f183a3c7d7a.png)



QUESTIONS:
1. How much is there real cases (without modified xud) when cancellation from counterparty can be reached?
2. Should cancellation work in third case for four transfers from third point (with incoming error)?
3. Why i got only 1 pending htlc but 2 monitored swaps?

sangaman commented:
Thank you @erkarl and @raladev for the tests, logs, and feedback. I think I figured out the issue which is that we were deleting the `expectedIncomingTransfer` (equivalent to an invoice) upon accepting the htlc, and then expecting it to be there when we go to cancel that htlc. That would result in the "could not find expected incoming transfer" log message and no invoice being canceled. I fixed this and refactored the code around it silghtly, there's no longer separate maps for expected transfer and paymentIds but now the expected transfers include the paymentId once the htlc is accepted. Could you please try this again and see if it's working as expected?

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -399,10 +400,26 @@ class ConnextClient extends SwapClient {
+        await this.sendRequest('/hashlock-resolve', 'POST', {
+          paymentId,
+          assetId: expectedIncomingTransfer.tokenAddress,
+        });
```
1 minor thing - can we add logging here so that it's analogical to `LndClient.removeInvoice`:
```
this.logger.debug(`canceled invoice for ${rHash}`);
```
